### PR TITLE
Add IrrLicht-specific smart pointer

### PR DESCRIPTION
**Pull request by numberZero**

### CODE REVIEW

**Comment by rubenwardy** [comment]
Am I right in thinking that this will allow us to stop manually using ->get and ->drop?

**Comment by numberZero** [comment]
@rubenwardy That is the intention. (I assume you meant `grab` and not `get`)

File: src/irr_ptr.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	}
+
+	~irr_ptr()
+	{
+		reset();
+	}
+
+	template <typename B, class = typename std::enable_if <
+		std::is_convertible <B *, ReferenceCounted *>::value>::type>
+	irr_ptr &operator= (const irr_ptr<B> &b) noexcept
```
Comment (juhdanad) [inline src/irr_ptr.h]: Is that space after the `=` sign intentional?

File: src/irr_ptr.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	{
+		reset(b.get());
+		return *this;
+	}
+
+	template <typename B, class = typename std::enable_if <
+		std::is_convertible <B *, ReferenceCounted *>::value>::type>
+	irr_ptr &operator= (irr_ptr<B> &&b) noexcept
```
Comment (juhdanad) [inline src/irr_ptr.h]: Same space here.

File: src/irr_ptr.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
+#pragma once
```
Comment (juhdanad) [inline src/irr_ptr.h]: I think you should add a license header.

File: src/irr_ptr.h (hunk 1 above)
Comment (numberZero) [reply src/irr_ptr.h]: Yes. Does that look bad? Operator overloading is tricky to format well. `void operator()() {}` is perfectly valid.

Comment (juhdanad) [reply src/irr_ptr.h]: There is no official rule that I know of, but if you watch `operator=` as a function name, no spaces are required. `clang-format` wants to remove the space.

**Comment by rubenwardy** [comment]
@JDCodeIt #7235

File: src/irr_ptr.h (hunk 3 above)
Comment (NewbProgrammer101) [reply src/irr_ptr.h]: ^

Comment (numberZero) [reply src/irr_ptr.h]: Done, in 913fe82

**Comment by numberZero** [comment]
@NewbProgrammer101 I could do that, would that be possible. But templates just don’t work that way. Also the template here is tiny, so compilation time is small; and the code may be called *very* often, so inline functions are preferable anyway.

Jay, template is not something that can be compiled and stored in an object file, like a function. It is just how it is called, function (or class) *template*. When you use it at some point with some specific parameters (even if you don’t supply them, as the compiler deduces them automatically in many cases), the compiler *instantiates* it, that is, builds a concrete function out of the template. But to do that, it needs the template source, and it needs the parameters *at the same time*.

In fact, template instantiation in one file may be used in any other, just like normal function. So if you know all the parameter sets template may ever be used with, you can move its code in a separate source file and instantiate it explicitly *for each parameter set*. Then the template will be usable exactly for these parameter sets. As you can see, such usage reduces flexibility a lot, but has some legitimate uses (e.g. if all parameters are enumerations only, you can just instantiate all possible combinations explicitly).

**Comment by numberZero** [comment]
One problem was pointed out recently: objects created with `add` still need deleting, but by means different from `drop`. So smart pointer should not probably be used for these at all, as it can’t delete them properly anyway. The `do_grab` option was intended specifically for such objects initially, but raw pointers are probably better for them.

**Comment by numberZero** [comment]
Updated. Tests *are* important.

**Pull request merged**